modules/heat_conduction/src/materials/FunctionPathEllipsoidHeatSource.C

```diff
@@ -0,0 +1,68 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "FunctionPathEllipsoidHeatSource.h"
+
+registerMooseObject("HeatConductionApp", FunctionPathEllipsoidHeatSource);
+
+template <>
+InputParameters
+validParams<FunctionPathEllipsoidHeatSource>()
+{
+  InputParameters params = Material::validParams();
+  params.addRequiredParam<Real>("power", "laser power");
+  params.addParam<Real>("efficiency", 1, "process efficiency");
+  params.addRequiredParam<Real>("rx", "transverse ellipsoid axe");
+  params.addRequiredParam<Real>("ry", "longitudinal ellipsoid axe");
+  params.addRequiredParam<Real>("rz", "depth ellipsoid axe");
+  params.addParam<Real>("factor", 1, "scaling factor");
+  params.addParam<FunctionName>("function_x", "0", "The x component heating spot travel path");
+  params.addParam<FunctionName>("function_y", "0", "The y component heating spot travel path");
+  params.addParam<FunctionName>("function_z", "0", "The z component heating spot travel path");
+  params.addClassDescription("Double ellipsoid volumetric source heat with function path.");
+
+  return params;
+}
+
+FunctionPathEllipsoidHeatSource::FunctionPathEllipsoidHeatSource(const InputParameters & parameters)
+  : Material(parameters),
+    _P(getParam<Real>("power")),
+    _eta(getParam<Real>("efficiency")),
+    _rx(getParam<Real>("rx")),
+    _ry(getParam<Real>("ry")),
+    _rz(getParam<Real>("rz")),
+    _f(getParam<Real>("factor")),
+    _function_x(isParamValid("function_x") ? &getFunction("function_x") : nullptr),
+    _function_y(isParamValid("function_y") ? &getFunction("function_y") : nullptr),
```

Comment:
I think because you have a default of 0 in these parameters, getFunction() will give you the 0 function by default if the user doesn't specify it, so you can just grab a reference to it and use it rather than having it be a pointer and setting it to nullptr if it's not defined.